ddtrace/contrib/cassandra/session.py

```diff
@@ -31,32 +36,133 @@ def traced_connect(func, instance, args, kwargs):
     session = func(*args, **kwargs)
     if not isinstance(session.execute, wrapt.FunctionWrapper):
         # FIXME[matt] this should probably be private.
-        setattr(session, 'execute', wrapt.FunctionWrapper(session.execute, traced_execute))
+        setattr(session, 'execute_async', wrapt.FunctionWrapper(session.execute_async, traced_execute_async))
     return session
 
-def traced_execute(func, instance, args, kwargs):
-    cluster = getattr(instance, 'cluster', None)
+def _close_span_on_success(result, future):
+    span = getattr(future, CURRENT_SPAN, None)
+    if not span:
+        log.debug('traced_set_final_result was not able to get the current span from the ResponseFuture')
+        return
+    with span:
+        span.set_tags(_extract_result_metas(cassandra.cluster.ResultSet(future, result)))
+
+def traced_set_final_result(func, instance, args, kwargs):
+    result = kwargs.get('result') or args[0]
+    _close_span_on_success(result, instance)
+    return func(*args, **kwargs)
+
+def _close_span_on_error(exc, future):
+    span = getattr(future, CURRENT_SPAN, None)
+    if not span:
+        log.debug('traced_set_final_exception was not able to get the current span from the ResponseFuture')
+        return
+    with span:
+        # FIXME how should we handle an exception that hasn't be rethrown yet
+        try:
+            raise exc
+        except:
+            span.set_exc_info(*sys.exc_info())
+
+def traced_set_final_exception(func, instance, args, kwargs):
+    exc = kwargs.get('result') or args[0]
+    _close_span_on_error(exc, instance)
+    return func(*args, **kwargs)
+
+def traced_start_fetching_next_page(func, instance, args, kwargs):
+    has_more_pages = instance._paging_state is not None
+    if not has_more_pages:
+        return func(*args, **kwargs)
+    session = getattr(instance, 'session', None)
+    cluster = getattr(session, 'cluster', None)
     pin = Pin.get_from(cluster)
     if not pin or not pin.enabled():
         return func(*args, **kwargs)
 
-    service = pin.service
-    tracer = pin.tracer
+    # In case the current span is not finished we make sure to finish it
```

Comment:
This could happen quite often? Is it anyway an expected behavior in the `execute_async` implementation?

Reply:
This should not happen if the ResponseFuture is used properly. But if someone where to call the `start_fetching_next_page` before the previous call is finished I don't want us to keep spans open.

Reply:
In general, if it could happen at most once, I'd prefer to have a safe-guard in any case. Keeping opened spans is something that leads to definitely wrong traces. So good to me.